Refactor: Add Flags enum to ArgsManager class

hebasto opened the pull request:
This PR adds the `Flags` enum to the `ArgsManager` class. Also the `m_flags` member is added to the `Arg` struct. Flags denote an allowed type of an arg value and special hints.

This PR is only a refactoring and does not change behavior.

ryanofsky commented on src/util/system.cpp:
```diff
@@ -319,6 +318,27 @@ ArgsManager::ArgsManager() :
       "-wallet",
+    },
+
+    m_allowed_negated_args{
```
The problem described in the PR is real but I don't think the change implemented is the simplest or best fix. Having a new whitelist of options located in a separate part of the code from where the options are used and defined adds complexity, inconsistency, and [action at distance](https://en.wikipedia.org/wiki/Action_at_a_distance_(computer_programming)) to already very confusing code.

If someone specifies `-nopid`, it is wrong to create a pid file called `0`, but a simpler and more useful fix would be to be not write a pid file instead of adding a new logic in a different part of the code to print a "negating not allowed" error.

But the `-pid` argument case actually seems atypical here. More cases where negated arguments are handled incorrectly (like `-incrementalrelayfee`) seem like simple misuses of the IsArgSet function, which is misnamed and doesn't do what people think it does. I'd replace the `IsArgSet` function with separate `ArgHasValue` and  `ArgIsSpecified` functions and use `ArgHasValue` instead of `ArgIsSpecified` everywhere or almost everywhere the code currently using ArgIsSet.

hebasto replied:
`m_allowed_negated_args` is used in the same manner as `m_network_only_args` (see #11862) and it seems consistent, IMO.

promag replied:
> adds complexity, inconsistency, and action at distance to already very confusing code

Agree with @ryanofsky. I'd rather see an approach like https://github.com/bitcoin/bitcoin/pull/16416#issuecomment-514120375

Beside the fact that `m_network_only_args` is very small, I think it could be improved like my comment above - move that flag to the `Arg` structure.

Currently `ArgsManager::Arg` has:
```cpp
        bool m_debug_only;
```
And I think it could be changed to something like:
```cpp
   enum Flags {
      NONE = 0x00,
      DEBUG_ONLY = 0x01,
      NETWORK_ONLY = 0x2,
      ALLOW_NEGATED = 0x04,      
   };

   Flags m_flags;
```

hebasto commented:
@ryanofsky Thank you for your review.

> If someone specifies `-nopid`, it is wrong to create a pid file called `0`, but a simpler and more useful fix would be to be not write a pid file instead of adding a new logic in a different part of the code to print a "negating not allowed" error.
> But the `-pid` argument case actually seems atypical here.

Agree that `-nopid` should have a special behavior. And it deserves a separate PR.

> More cases where negated arguments are handled incorrectly (like `-incrementalrelayfee`) seem like simple misuses of the IsArgSet function, which is misnamed and doesn't do what people think it does. I'd replace the `IsArgSet` function with separate `ArgHasValue` and `ArgIsSpecified` functions and use `ArgHasValue` instead of `ArgIsSpecified` everywhere or almost everywhere the code currently using ArgIsSet.

I don't think this approach will compeltely fix the problem described in this PR.

hebasto commented:
@promag 

> And I think it could be changed to something like:
```
   enum Flags {
      NONE = 0x00,
      DEBUG_ONLY = 0x01,
      NETWORK_ONLY = 0x2,
      ALLOW_NEGATED = 0x04,      
   };

   Flags m_flags;
```

Done.

hebasto commented:
Commit 50a7f97db2a9daa1ed4f32e64cfb87b02e205e70 needs rebase on top of #16366. Going to make it tonight.

ryanofsky commented:
Thanks for the update! I still see several problems with this approach, but I think they can be fixed and I have specific suggestions below that I think will put this PR in a good state and make it easier to review. Here are the problems that I see presently (90dfd237ea75301891fc1715ebf319966f2e0b86):

- Interaction between boolean option detection via help string and ALLOW_NEGATED errors via flags is confusing and I think will lead to bugs in the future.
- PR has no [release notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#release-notes), no clear description of changes in behavior understandable to a normal user.
- PR mixes refactoring changes with behavioral changes, and makes behavior changes in commits that affects swaths of options rather than individual options, so it is difficult as a reviewer to determine how behavior is changing and verify the changes are good without searching and looking at every option.

I think all of these problems are fixable. Here is are my suggestions:

1. Drop the detection of boolean arguments through help strings and drop the `ALLOW_NEGATED` flag. The only flags I would suggest keeping here are:

```c++
ALLOW_BOOL = 0x01,
ALLOW_INT = 0x02,
ALLOW_STRING = 0x4,
ALLOW_ANY = ALLOW_BOOL | ALLOW_INT | ALLOW_STRING,
DEBUG_ONLY = 0x100,
NETWORK_ONLY = 0x200,
```

2. Do not change any behavior in this PR. Set every existing option to `ALLOW_ANY` by default, with `| DEBUG_ONLY` and `| NETWORK_ONLY` flags added as necessary to preserve behavior.

3. Raise an error during startup if a `-noXXX` option is given and the `ALLOW_BOOL` flag is not present. This should have no effect on behavior of existing options because `ALLOW_ANY` should be set as described above.

4. Optionally, if you feel like implementing extra error checking that could be useful in the future, check for properly formatted ints and bools and nonempty strings based on the int/bool/string flags. Again these checks should not change behavior of existing options because `ALLOW_ANY` should always be set here.

5. In a followup PR, replace `ALLOW_ANY` with `ALLOW_STRING` or `ALLOW_INT` for selected options where the goal is to disallow negation. This PR should be small, trivial to review, and have [release notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#release-notes) describing the user visible changes and options that are affected.

6. Make separate PRs for help text cleanups, any other unrelated cleanups.

Overall, I think this is moving in a good direction and adding the flags really helps. But the more we can move in the direction of simple bool/int/string option types, and have negation taken care of on the front end, rather than leaking into entire codebase, the better state this code will be in, and less weird behavior we will have in the future.

hebasto commented:
@ryanofsky Thank you for your review. I really appreciate it.

> 5. In a followup PR, replace ALLOW_ANY with ALLOW_STRING or ALLOW_INT for selected options where the goal is to disallow negation. This PR should be small, trivial to review, and have release notes describing the user visible changes and options that are affected.

We have over two hundred target lines of code. Proposed approach does not allow to apply a scripted-diff, unfortunately. Thoughts?

hebasto commented:
@ryanofsky 
> Drop the detection of boolean arguments through help strings ...

Agree. Going to implement it.

> ... and drop the `ALLOW_NEGATED` flag.

I'd prefer to use it. It complements the `IsArgNegated()` function. Also, consider the following two options:
- `-nowallet` has a special meaning, `-wallet=0` points to a wallet named `0` and `-nowallet=0` is an error; possible flags are `ALLOW_BOOL | ALLOW_STRING`
- `-noprune` equals to `-prune=0`; `-noprune=0` is not an error; possible flags are `ALLOW_BOOL | ALLOW_INT`

Using the `ALLOW_NEGATED` flag makes things simpler, IMO:
- `-wallet` has flags `ALLOW_STRING | ALLOW_NEGATED`
- `-prune` has flags `ALLOW_INT | ALLOW_NEGATED`

ryanofsky commented:
> > 5. In a followup PR, replace ALLOW_ANY with ALLOW_STRING or ALLOW_INT for selected options where the goal is to disallow negation. This PR should be small, trivial to review, and have release notes describing the user visible changes and options that are affected.
> 
> We have over two hundred target lines of code. Proposed approach does not allow to apply a scripted-diff, unfortunately. Thoughts?

That's the whole point! It's good to use scripted diffs in refactoring commits, but there should be no scripted diff for the behavior change. The code changes in the behavior change PR should just look like:

```diff
-gArgs.AddArg("-changetype=<n>", ALLOW_ANY, ...);
+gArgs.AddArg("-changetype=<n>", ALLOW_STRING, ...);
```

```diff
-gArgs.AddArg("-dbcache=<n>", ALLOW_ANY, ...);
+gArgs.AddArg("-dbcache=<n>", ALLOW_INT, ...);
```

So it is easy for reviewers to verify that new errors triggered on options that currently allow negation are appropriate and appropriately documented in release notes.

There should not be 200 lines targeted. Every existing boolean option, every existing list option, most string options, and many int options that accept 0 should continue to allow negation.

> > ... and drop the `ALLOW_NEGATED` flag.
> 
> I'd prefer to use it. It complements the `IsArgNegated()` function.

Just like _IsArgSet_ which is misused many places, misunderstood and should be eliminated. 
_IsArgNegated_ is an awkward API that should go away, and `ALLOW_NEGATED` should never be introduced.

If you look at [google flags api](https://abseil.io/docs/python/guides/flags), [qsettings](https://doc.qt.io/qt-5/qsettings.html), [python argparse](https://docs.python.org/3/library/argparse.html), the standard everywhere is bool/number/string representations, not "is negated" queries for code reading the values. My refactoring PR #15934 is a major step in this direction internally replacing our vector representation of negated values with plain, typed `false` values.

Nothing I'm suggesting here involves changing or limiting behavior in any way. I'm just suggesting to use normal, sane terminology, so we have a more usable api and so the IsArgSet IsArgNegated bugs you are fixing now don't get reintroduced in the future.
 
> Using the `ALLOW_NEGATED` flag makes things simpler, IMO:
> 
> * `-wallet` has flags `ALLOW_STRING | ALLOW_NEGATED`
> * `-prune` has flags `ALLOW_INT | ALLOW_NEGATED`

That's exactly what I'm suggesting, except spelling `ALLOW_NEGATED` as  `ALLOW_BOOL`.

I didn't fully describe in my previous comment how the suggested flags should behave, so I'll do that now to avoid unnecessary confusion. I don't think the extra error checks I'm describing here need to be implemented now. It'd be fine to implement them in future PRs that actually start adding `ALLOW_BOOL` / `ALLOW_INT` / `ALLOW_STRING` options instead of `ALLOW_ANY` for everything. Anyway, here's how the flags are meant to behave:

* `-noOPTION` and `-noOPTION=1` negation syntax should be accepted when `ALLOW_BOOL` is present.
* `-OPTION` and `-OPTION=` empty string syntax be accepted when `ALLOW_BOOL` is present.
* `-OPTION=0` and `-OPTION=1` syntax should be accepted when `ALLOW_BOOL` is present.
* `-OPTION=123` where _123_ is an integer should be accepted when `ALLOW_INT` is present.
* `-OPTION=abc` where _abc_ is a nonempty string should be accepted when `ALLOW_STRING` is present.
* Anything not accepted above should trigger an error.

Except for one detail, adding these checks should have no effect when ALLOW_ANY is specified, so even if these checks are added now, this can remain a pure refactoring PR and not change behavior. 

The detail is that we currently accept `-noOPTION=abc` syntax for strange values of _abc_, sometimes interpreting _abc_ as false and showing a warning, sometimes interpreting it as true and not warning. This behavior should be left untouched here, and cleaned up in a separate PR (making any _abc_ that is not the empty string or 1 an error).

promag commented:
Really nice @ryanofsky. Maybe it could also have a `MULTIPLE` flag?

ryanofsky commented:
> Maybe it could also have a MULTIPLE flag?

Yes I think that'd be a good flag, and I also like your idea adding support to declare default values. Best if this PR scope doesn't expand too much, but going forward, the more things about options that can be declared up front, the less complicated code reading options has to be and the less chance for inconsistencies and bugs.

promag commented:
> Best if this PR scope doesn't expand too much

👍 

Another flag (otherwise I may forget it) could be `SENSITIVE` so that its value wouldn't be dumped.

hebasto commented:
@ryanofsky 
>     6. Make separate PRs for help text cleanups, any other unrelated cleanups.

The "_refactoring: Use direct list initialization_" commit (b89eb79b3a7bc3c2a0cd80454f4373f7a07f2fd5) has split out to #16469.

ryanofsky commented:
> @ryanofsky
> 
> > ```
> > 6. Make separate PRs for help text cleanups, any other unrelated cleanups.
> > ```
> 
> The "_refactoring: Use direct list initialization_" commit ([b89eb79](https://github.com/bitcoin/bitcoin/commit/b89eb79b3a7bc3c2a0cd80454f4373f7a07f2fd5)) has split out to #16469.

Sorry, I'd suggest closing #16469. The quote above is a little ambiguous, but I was trying to suggest splitting **behavior** changes into a separate PRs from the requested flags refactor, not creating multiple refactoring PRs.

It's good to make behavior changes in small independent PRs so they don't get missed by reviewers and so there aren't multiple confusing discussion about different behaviors. It generally takes a lot less mental effort to review a pure refactoring PR than it does to review a refactoring PR mixed with behavior changes, so there shouldn't be any problem with including a small list initialization refactor in the same PR that does the requested flag refactor.

MarcoFalke commented:
Agree with @ryanofsky

hebasto committed:
refactoring: Check IsArgKnown() early

hebasto committed:
Refactor InterpretNegatedOption() function

- added args parameter
- renamed to InterpretOption()
- removed code duplication

hebasto committed:
Add Flags enum to ArgsManager

hebasto committed:
scripted-diff: Use Flags enum in AddArg()

-BEGIN VERIFY SCRIPT-
sed -i 's/const bool debug_only,/unsigned int flags, &/' src/util/system.h src/util/system.cpp
sed -i -E 's/(true|false), OptionsCategory::/ArgsManager::ALLOW_ANY, &/' $(git grep --files-with-matches 'AddArg(' src)
-END VERIFY SCRIPT-

hebasto committed:
scripted-diff: Use ArgsManager::DEBUG_ONLY flag

-BEGIN VERIFY SCRIPT-
sed -i 's/unsigned int flags, const bool debug_only,/unsigned int flags,/' src/util/system.h src/util/system.cpp
sed -i 's/ArgsManager::NONE, debug_only/flags, false/' src/util/system.cpp
sed -i 's/arg.second.m_debug_only/(arg.second.m_flags \& ArgsManager::DEBUG_ONLY)/' src/util/system.cpp
sed -i 's/ArgsManager::ALLOW_ANY, true, OptionsCategory::/ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::/' $(git grep --files-with-matches 'AddArg(' src)
sed -i 's/ArgsManager::ALLOW_ANY, false, OptionsCategory::/ArgsManager::ALLOW_ANY, OptionsCategory::/' $(git grep --files-with-matches 'AddArg(' src)
-END VERIFY SCRIPT-

hebasto committed:
Remove unused m_debug_only member from Arg struct

hebasto committed:
Use ArgsManager::NETWORK_ONLY flag

hebasto committed:
Replace IsArgKnown() with FlagsOfKnownArg()

hebasto committed:
Revamp option negating policy

hebasto committed:
test: Make tests arg type specific

hebasto commented:
@ryanofsky 
> 2. Do not change any behavior in this PR. Set every existing option to `ALLOW_ANY` by default, with `| DEBUG_ONLY` and `| NETWORK_ONLY` flags added as necessary to preserve behavior.
> 3. Raise an error during startup if a `-noXXX` option is given and the `ALLOW_BOOL` flag is not present. This should have no effect on behavior of existing options because `ALLOW_ANY` should be set as described above.

Done.

> 5. In a followup PR, replace `ALLOW_ANY` with `ALLOW_STRING` or `ALLOW_INT` for selected options where the goal is to disallow negation. This PR should be small, trivial to review, and have release notes describing the user visible changes and options that are affected.

`ALLOW_ANY` is replaced with `ALLOW_STRING` in #16476.

hebasto commented:
PR title and description have been updated.

ryanofsky commented on test/functional/feature_config_args.py:
```diff
@@ -78,7 +78,7 @@ def test_config_file_parser(self):
 
     def test_log_buffer(self):
-        with self.nodes[0].assert_debug_log(expected_msgs=['Warning: parsed potentially confusing double-negative -connect=0']):
+        with self.nodes[0].assert_debug_log(expected_msgs=['Warning: parsed potentially confusing double-negative -noconnect=0']):
```
In commit "Revamp option negating policy" (3f0b5e28ebcf42c168f393a3109403354cb8e128)

This commit appears to be changing behavior by fixing the error message so it technically contradicts "This PR is only a refactoring and does not change behavior" in the PR description. Could consider removing this here and just doing it as a minor bugfix later. Or updating the PR description to mention the change in behavior.

hebasto replied:
> Could consider removing this here and just doing it as a minor bugfix later.

Done.

ryanofsky reviewed (commented):
This looks great! I need to check this more throughly to give a review ACK, but I went through all the commits, and they are all simple, well structured cleanups that should make this easy to merge.

jamesob commented on src/util/system.cpp:
```diff
@@ -294,17 +293,18 @@ static bool InterpretNegatedOption(std::string& key, std::string& val)
         key.erase(option_index, 2);
         if (!bool_val ) {
             // Double negatives like -nofoo=0 are supported (but discouraged)
             LogPrintf("Warning: parsed potentially confusing double-negative %s=%s\n", key, val);
```
(https://github.com/bitcoin/bitcoin/pull/16097/commits/e0d187dfeb18b026de22bd7960b2a50c2b958e1a)
 
FWIW, this results in a slightly incorrect error message:
```
$ ./src/bitcoind -datadir=/data/bitcoin -noconnect=0 -logthreadnames

2019-08-01T16:03:23Z Warning: parsed potentially confusing double-negative -connect=0
```
which I think should read "double-negative -noconnect=0".

hebasto replied:
> I think should read "double-negative -noconnect=0".

You are right. This remained untouched to preserve the old behavior and make this PR refactoring only. Details: https://github.com/bitcoin/bitcoin/pull/16097#discussion_r308829507 and https://github.com/bitcoin/bitcoin/pull/16097#discussion_r309281546.

jamesob reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/16097/commits/e6f649cb2c07bf55d9214c2876619c56f1d6fe30

### Commit checklist

- [x] e0e18a1017 refactoring: Check IsArgKnown() early

  Checks whether an argument is recognized before determining whether it's a
  negation.

- [x] e0d187dfeb Refactor InterpretNegatedOption() function

  Changes the way negated options are processed (DRYs up code, modifies args
  map inline).

- [x] 265c1b58d8 Add Flags enum to ArgsManager

  Defines the various flags available to ArgsManager.

- [x] 1b4b9422ca scripted-diff: Use Flags enum in AddArg()

  Straightforward scripted-diff adding `flags` argument to AddArg(), filling
  with degenerate ALLOW_ANY default.

- [x] fb4b9f9e3b scripted-diff: Use ArgsManager::DEBUG_ONLY flag

  Collapses the AddArg `debug_only` parameter into `flags`. Cumbersome to
  review even for a scripted-diff because of the remarkable line length and
  number of changes but... it's a scripted-diff on arg processing code ¯\_(ツ)_/¯.

- [x] 9a12733508 Remove unused m_debug_only member from Arg struct

- [x] dde80c272a Use ArgsManager::NETWORK_ONLY flag

  Manually specifies NETWORK_ONLY flag for certain arguments that were
  previously special-cased away from their AddArg definitions.

- [x] db08edb303 Replace IsArgKnown() with FlagsOfKnownArg()
- [x] b70cc5d733 Revamp option negating policy
- [x] e6f649cb2c test: Make tests arg type specific


### Tests

Built and ran `./src/test/test_bitcoin` successfully.

Not sure if this was a preexisting issue, but it's kind of weird that when I
double-negate `connect`, we seem to get socket errors.

```
./src/bitcoind -datadir=/data/bitcoin -noconnect=0 -logthreadnames

2019-08-01T16:03:23Z Warning: parsed potentially confusing double-negative -connect=0

...

2019-08-01T16:03:40Z [opencon] opencon thread start
2019-08-01T16:03:40Z [msghand] msghand thread start
2019-08-01T16:03:40Z [opencon] socket send error Broken pipe (32)
2019-08-01T16:03:40Z [opencon] socket send error Broken pipe (32)
2019-08-01T16:03:41Z [opencon] socket send error Broken pipe (32)
2019-08-01T16:03:43Z [opencon] socket send error Broken pipe (32)
2019-08-01T16:03:45Z [opencon] socket send error Broken pipe (32)
2019-08-01T16:03:47Z [opencon] socket send error Broken pipe (32)
2019-08-01T16:03:50Z [opencon] socket send error Broken pipe (32)
2019-08-01T16:03:54Z [opencon] socket send error Broken pipe (32)
```

I'm not really sure how to test the `NETWORK_ONLY` stuff (which doesn't seem to
be covered by unittests), but this works:

```
$ ./src/bitcoind -datadir=/data/bitcoin -testnet -port=4444 -nologthreadnames=0

[as expected]
```

ryanofsky commented:
I'm glad this PR isn't trying to fully implement this flags yet, since it keeps the change light and simple, and because the flag implementations will make more sense in context of followup PRs which actually put them to use. But I think I would like to see simple asserts added here to prevent misapplication of the new flags by developers, and bugs in the future when the flags actually do start being enforced:

```diff
diff --git a/src/util/system.cpp b/src/util/system.cpp
index 5f704ecabef..8ad02128c57 100644
--- a/src/util/system.cpp
+++ b/src/util/system.cpp
@@ -461,6 +461,7 @@ unsigned int ArgsManager::FlagsOfKnownArg(const std::string& key) const
 
 std::vector<std::string> ArgsManager::GetArgs(const std::string& strArg) const
 {
+    assert(FlagsOfKnownArg(strArg) & ALLOW_STRING);
     std::vector<std::string> result = {};
     if (IsArgNegated(strArg)) return result; // special case
 
@@ -504,6 +505,7 @@ bool ArgsManager::IsArgNegated(const std::string& strArg) const
 
 std::string ArgsManager::GetArg(const std::string& strArg, const std::string& strDefault) const
 {
+    assert(FlagsOfKnownArg(strArg) & ALLOW_STRING);
     if (IsArgNegated(strArg)) return "0";
     std::pair<bool,std::string> found_res = ArgsManagerHelper::GetArg(*this, strArg);
     if (found_res.first) return found_res.second;
@@ -512,6 +514,7 @@ std::string ArgsManager::GetArg(const std::string& strArg, const std::string& st
 
 int64_t ArgsManager::GetArg(const std::string& strArg, int64_t nDefault) const
 {
+    assert(FlagsOfKnownArg(strArg) & ALLOW_INT);
     if (IsArgNegated(strArg)) return 0;
     std::pair<bool,std::string> found_res = ArgsManagerHelper::GetArg(*this, strArg);
     if (found_res.first) return atoi64(found_res.second);
@@ -520,6 +523,7 @@ int64_t ArgsManager::GetArg(const std::string& strArg, int64_t nDefault) const
 
 bool ArgsManager::GetBoolArg(const std::string& strArg, bool fDefault) const
 {
+    assert(FlagsOfKnownArg(strArg) & ALLOW_BOOL);
     if (IsArgNegated(strArg)) return false;
     std::pair<bool,std::string> found_res = ArgsManagerHelper::GetArg(*this, strArg);
     if (found_res.first) return InterpretBool(found_res.second);
```

maflcko reviewed (approved):
ACK e6f649cb2c07bf55d9214c2876619c56f1d6fe30 thanks for adding types to the command line options

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK e6f649cb2c07bf55d9214c2876619c56f1d6fe30 thanks for adding types to the command line options
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYEACgkQzit1aX5p
pUixQAv+PcIolVlZDyU/bWh0rKnyH4rRnFKm/ZrjcLrXWFby/fC5x8spb/1Kizcx
TtMKnLEP3YXZTo0PgOyqIemGsVit+6NkIVvpIbPjz/M6VTz1CLLZp0rdRWAHxUKs
hX3Un0yAch+IK/IEq3J813TOSCFlHWkdy+5LNzC77sGx5M4nJk302QmsRvlAgEsc
IlJAfSqEk1wF6MaQ9yhdIrnfOBiiM/ZnXCOQK6VAO4uw2xUrl1Cd3f9L8lO8CxWm
O+3vy51F84af2oDSpKuhmT4Vcyl+/H25UNb3pNWRky+KmmGYSGKXSuWWOz8EK9/3
KHq6pEajoRWDG6qsZnUyzS4QQAx9Vf2ZTsBh37UmRj3AgS6Ct/jKsYJvL69QP+C4
9zPIs4NQNP24dkGTNx7R7TcoUU0P60O72Mop8aqEhwnQPaffyXlWTuRs8sjYdsBL
62JcofhWZRG++V8l9bafcyBKCn6sIV0C6ztAgDgvR0gW7wrX0xOdckUyrRecDncP
U47QO/n4
=pQlB
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `60ba83da959664c8ccc4f7e6dac1c5da28003ce06779868d2def12da094a0be4  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010860ba83da959664c8ccc4f7e6dac1c5da28003ce06779868d2def12da094a0be4f0106b09262cd69cc5cecda6d9e57dee2a4b08fff01003735644fdcc07462165ac54e7a4c38508f1045d446219f008b96a5fcd637c007c0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010b37797e29aa90739a5c168cd471cac8508f1045d446219f00812e8b67153e5c73b0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010b606c537ad0851aa773509749391d32008f1045d446219f008829b6c00fd0f76770083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010e3e55ef4d0d7f39db23ada0832f846df08f1045d446218f0086939d476e313c05d0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

ryanofsky reviewed (commented):
utACK e6f649cb2c07bf55d9214c2876619c56f1d6fe30 (post-merge)
